program-runtime/src/loaded_programs.rs

```diff
@@ -258,11 +287,18 @@ impl LoadedPrograms {
         let index = second_level
             .iter()
             .position(|at| at.effective_slot >= entry.effective_slot);
-        if let Some(existing) = index.and_then(|index| second_level.get(index)) {
+        if let Some((existing, entry_index)) =
+            index.and_then(|index| second_level.get(index).map(|value| (value, index)))
+        {
             if existing.deployment_slot == entry.deployment_slot
                 && existing.effective_slot == entry.effective_slot
             {
-                return (true, existing.clone());
+                if matches!(existing.program, LoadedProgramType::Unloaded) {
+                    // The unloaded program is getting reloaded
+                    second_level.swap_remove(entry_index);
```

Comment:
Doesn't this delete the statistics accumulated in the entry so far?

Reply:
Thanks, that's a good catch. Updated the PR to copy over the usage counter before removing the entry.